### PR TITLE
added translate, scale, their test files, and docs

### LINKED ISSUE #88: Add translate and scale function
Contact @lawruble13 (Liam Wrubleski) if you need help/guidance

Expose the the `translate()` and the `scale()` functions from the ipycanvas library so that students can use it like regular functions. See Issue #86 for the `rotate()` function.
ipycanvas docs for functions found here: https://ipycanvas.readthedocs.io/en/latest/transformations.html

Remember to create a Pull Request to merge your code from your branch into the `dev` branch and notify the Library team on Slack

AC:
* [x] Expose `translate(x, y)`
* [x] Expose `scale(x, y=none)`
* [x] Add unit tests
* [x] Contact @Descent098 to learn how to add documentation

### PR DESCRIPTION
**Pull request by VB256**
Closes #88

### CODE REVIEW

File: docs/utilities.md [hunk 1]
```diff
@@ -124,4 +124,105 @@ def setup():
 
 Results in:
 
-![dist demo](img/dist_2.png)
\ No newline at end of file
+![dist demo](img/dist_2.png)
+
+### Scale (one parameter)
```
Comment (AlphaRLee) [inline docs/utilities.md]: Inconsistent with the rest of our docs. Can you merge your `Scale (one parameter)` and `Scale (two parameters)` docs into one single section? Take a look at the docs for [colors](https://spark.schulichignite.com/color/#changing-the-color-of-shapes) on our site for examples on how to do this
```suggestion
### Scale
```

File: docs/utilities.md [hunk 2]
```diff
@@ -124,4 +124,105 @@ def setup():
...
+
+Scales the canvas units by x horizontally and vertically.
```
Comment (AlphaRLee) [inline docs/utilities.md]: Avoid literally re-stating the code in the docs (just in English). Specifically, avoid mentioning the specific parameter names (unless necessary for understanding), that's what you have the `parameters` section for in the first place. In this case, you can just say "Scales the canvas", but it would be preferred to note that this transforms the entire canvas, not just one shape (see the notes on `### Rotate`)

File: docs/utilities.md [hunk 3]
```diff
@@ -124,4 +124,105 @@ def setup():
...
+```python
+scale(x)
+```
+**Parameters**
+
+- x: (float) The factor by which the canvas is scaled.
+
+**Example(s):**
+
+*Scale horizontal and vertical canvas units by a factor of 2*
```
Comment (AlphaRLee) [inline docs/utilities.md]: Nitpick: to me it's easier to understand if you say "doubles" instead of "scales by factor of two". Perks of giving an unambiguous example in case if people are confused by "The factor by which the canvas is scaled" and seeking an example
```suggestion
*Scale the width and height of the canvas units by double*
```

File: docs/utilities.md [hunk 4]
```diff
@@ -124,4 +124,105 @@ def setup():
...
+    # apply scale of 2, this will scale canvas units by factor of 2 horizontally and vertically
+    scale(2)
+    circle(100, 100, 100)
+```
+
+Results in:
+
+![scale demo](img/scale_1.png)
+
+### Scale (two parameters)
```
Comment (AlphaRLee) [inline docs/utilities.md]: See comments above, delete this section and move the example from here onto the end of the `### Scale` section

File: docs/utilities.md (hunk 1 above)
Comment (AlphaRLee) [reply docs/utilities.md]: Also move these entire sections (scale, translate) to be with the `### Rotate` docs. No point having them separated by `### dist`. 

Personal opinion, make the order in our docs:
1. Translate
2. Rotate
3. Scale

File: docs/utilities.md [hunk 5]
```diff
@@ -124,4 +124,105 @@ def setup():
...
+Scales the canvas units by x horizontally and vertically. 
+
+Usage:
+
+```python
+scale(x)
+```
+**Parameters**
+
+- x: (float) The factor by which the canvas is scaled.
```
Comment (AlphaRLee) [inline docs/utilities.md]: I recommend using a different variable (like `n`) to avoid confusion between `x` units being an arbitrary number vs `x` units being the horizontal transformation. Please also make appropriate changes through the rest of the docs
```suggestion
- n: (float) The factor by which the canvas is scaled.
```

File: docs/utilities.md [hunk 6]
```diff
@@ -124,4 +124,105 @@ def setup():
...
+    circle(100, 100, 100)
+```
+
+Results in:
+
+![scale demo](img/scale_2.png)
+
+### Translate
+
+Moves the canvas and its origin on the grid. x indicates the horizontal distance to move, and y indicates vertical distance to move.
```
Comment (AlphaRLee) [inline docs/utilities.md]: Don't mention the parameter names (see comments above)
Also I'm confused by "the canvas and its origin on the grid"
* Why separate "canvas and its origin", does this imply you can move the two separately through some other function?
* What's the definition of "the grid"? You can just omit this completely since you never need to reference this "grid" again (otherwise there's burden to give a proper definition)
```suggestion
Change the origin of the canvas.
```

File: docs/utilities.md [hunk 7]
```diff
@@ -124,4 +124,105 @@ def setup():
...
+
+Usage:
+
+```python
+translate(x,y)
```
Comment (AlphaRLee) [inline docs/utilities.md]: PEP8 convention: Add spaces after your commas
```suggestion
translate(x, y)
```

File: docs/utilities.md [hunk 8]
```diff
@@ -124,4 +124,105 @@ def setup():
...
+![scale demo](img/scale_1.png)
+
+### Scale (two parameters)
+
+Scales the canvas units by x horizontally and by y vertically. 
+
+Usage:
+
+```python
+scale(x,y)
```
Comment (AlphaRLee) [inline docs/utilities.md]: PEP8: Add spaces after your commas
```suggestion
scale(x, y)
```
If in doubt about style, then take a look at the [styling guide](https://docs.google.com/presentation/d/1o1pLhkoDgmy1QIT-70sZVw5N-t9IWiQOCMnyPp-gfhg/edit?usp=sharing) we use for our slide decks

File: docs/utilities.md [hunk 9]
```diff
@@ -124,4 +124,105 @@ def setup():
...
+
+- x: (float) The horizontal distance to translate the canvas.
+- y: (float) The vertical distance to translate the canvas.
+
+**Example(s):**
+
+*Translate the canvas 50 units right and 75 units down*
+
+```python hl_lines="6"
+%%ignite
```
Comment (AlphaRLee) [inline docs/utilities.md]:
```suggestion
```python hl_lines="9"
%%ignite

```

File: docs/utilities.md [hunk 10]
```diff
@@ -124,4 +124,105 @@ def setup():
...
+**Parameters**
+
+- x: (float) The factor by which the canvas is scaled.
+
+**Example(s):**
+
+*Scale horizontal and vertical canvas units by a factor of 2*
+
+```python hl_lines="6"
+%%ignite
```
Comment (AlphaRLee) [inline docs/utilities.md]: You're highlighting an empty line with this and missing a line break after `%%ignite`` (for consistency). Try 9?
```suggestion
```python hl_lines="9"
%%ignite

```

File: docs/utilities.md [hunk 11]
```diff
@@ -124,4 +124,105 @@ def setup():
...
+
+- x: (float) The factor by which the canvas is scaled horizontally.
+- y: (float) The factor by which the canvas is scaled vertically.
+
+**Example(s):**
+
+*Scale horizontal and vertical canvas units by factors of 0.75 and 1.25 respectively*
+
+```python hl_lines="6"
+%%ignite
```
Comment (AlphaRLee) [inline docs/utilities.md]: Again, change to 9 and add line break
```suggestion
```python hl_lines="9"
%%ignite

```

File: spark/core.py [hunk 12]
```diff
@@ -333,6 +333,12 @@ def background(self, *args): pass
 
     @extern
     def rotate(self, *args): pass
+    
+    @extern
+    def translate(self, *args): pass
+    
+    @extern
+    def scale(self, *args): pass
```
Comment (AlphaRLee) [inline spark/core.py]: I recommend flipping the order
```suggestion
    @extern
    def translate(self, *args): pass
    
    @extern
    def rotate(self, *args): pass
    
    @extern
    def scale(self, *args): pass
```

File: spark/util/helper_functions/canvas_functions.py [hunk 13]
```diff
@@ -40,3 +40,16 @@ def helper_background(self, *args):
...
+def helper_translate(self, *args):
+    self.canvas.translate(args[0], args[1])
+    
+@validate_args([Real], [Real, Real])
+@ignite_global
+def helper_scale(self, *args):
+    if len(args) == 1:
+        self.canvas.scale(args[0])
+    else:
+        self.canvas.scale(args[0], args[1])
```
Comment (AlphaRLee) [inline spark/util/helper_functions/canvas_functions.py]: You can condense this to a single line with list destructuring
```suggestion
    self.canvas.scale(*args)
```

File: docs/utilities.md [hunk 14]
```diff
@@ -124,4 +124,105 @@ def setup():
...
+
+**Example(s):**
+
+*Translate the canvas 50 units right and 75 units down*
+
+```python hl_lines="6"
+%%ignite
+def setup():
+    size(400, 400)
+    circle(100, 100, 100)
```
Comment (AlphaRLee) [inline docs/utilities.md]: This extra circle just complicates the example (remember to edit the highlight line `hl_lines` accordingly)
```suggestion
```

File: docs/utilities.md [hunk 15]
```diff
@@ -124,4 +124,105 @@ def setup():
...
+
+**Example(s):**
+
+*Scale horizontal and vertical canvas units by factors of 0.75 and 1.25 respectively*
+
+```python hl_lines="6"
+%%ignite
+def setup():
+    size(400, 400)
+    circle(100, 100, 100)
```
Comment (AlphaRLee) [inline docs/utilities.md]: This extra circle just complicates the example (remember to edit the highlight line `hl_lines` accordingly)
```suggestion
```

File: docs/utilities.md [hunk 16]
```diff
@@ -124,4 +124,105 @@ def setup():
...
+
+**Example(s):**
+
+*Scale horizontal and vertical canvas units by a factor of 2*
+
+```python hl_lines="6"
+%%ignite
+def setup():
+    size(400, 400)
+    circle(100, 100, 100)
```
Comment (AlphaRLee) [inline docs/utilities.md]: This extra circle just complicates the example. Also this one is completely hidden by the scaled up example... (remember to edit the highlight line `hl_lines` accordingly)
```suggestion
```

File: docs/utilities.md (hunk 1 above)
Comment (AlphaRLee) [reply docs/utilities.md]: Also preferred if you can quickly mention inside the `rotate` and `scale` descriptions that you can rotate or scale on a point other than the origin by first calling `translate` to change the origin to the new point to rotate from

File: test/ScaleTest.ipynb [hunk 17]
```diff
@@ -0,0 +1,1322 @@
+{
```
Comment (AlphaRLee) [inline test/ScaleTest.ipynb:1]: Can you add unit tests for:
1. Scaling one dimension by a factor of `0` (expected to not appear)
2. Scaling one dimension by a negative number. You'll probably need to translate to the middle of the screen first to see where the scaled result lands (expected to treat right as left or up as down and vice versa)

Comment (AlphaRLee) [reply test/ScaleTest.ipynb:1]: PS I like your unit tests for testing scaling by 1 and scaling only one dimension by 1 and another by 3 :+1:

File: test/TranslateTest.ipynb [hunk 18]
```diff
@@ -0,0 +1,1101 @@
...
+   ],
+   "source": [
+    "%%ignite\n",
+    "def setup():\n",
+    "    size(400, 400)\n",
+    "    circle(100, 100, 100)\n",
+    "    fill_style(\"red\")\n",
+    "    \n",
+    "    # move canvas 125 units left, and 10 units up\n",
+    "    translate(-25, -10)\n",
```
Comment (AlphaRLee) [inline test/TranslateTest.ipynb]: Typo, you said `125`, did you mean `25`?
```suggestion
    "    # move canvas 25 units left, and 10 units up\n",
    "    translate(-25, -10)\n",
```

File: test/TranslateTest.ipynb [hunk 19]
```diff
@@ -0,0 +1,1101 @@
...
+    "%%ignite\n",
+    "def setup():\n",
+    "    size(400, 400)\n",
+    "    circle(100, 100, 100)\n",
+    "    fill_style(\"red\")\n",
+    "    \n",
+    "    # move canvas 50 units right, and 75 units down\n",
+    "    translate(50, 75)\n",
+    "    circle(100, 100, 100)"
+   ]
```
Comment (AlphaRLee) [inline test/TranslateTest.ipynb]: Move this unit test to be the first one on the list (organize your unit tests so the most obvious "happy path" use case comes first and then all the obscure edge case ones follow)

File: test/TranslateTest.ipynb [hunk 20]
```diff
@@ -0,0 +1,1101 @@
...
+    "%%ignite\n",
+    "def setup():\n",
+    "    size(400, 400)\n",
+    "    circle(100, 100, 100)\n",
+    "    fill_style(\"red\")\n",
+    "    \n",
+    "    # move canvas 50 units left, and 80 units down\n",
+    "    translate(-50, 80)\n",
+    "    circle(100, 100, 100)"
+   ]
```
Comment (AlphaRLee) [inline test/TranslateTest.ipynb]: This is getting redundant, I think you already have a unit test doing this. Safe to delete?

**Review by AlphaRLee — CHANGES_REQUESTED** [review]
Code looks good but lots of nitpicks about your docs formatting and a couple edge cases missing from your scale unit tests

**Comment by AlphaRLee** [comment]
Also could you quickly try a unit test for `translate` with floats? Right now you have ints. Should work fine, but I don't want to take chances (libraries like Pygame crash when you feed floats)

**Review by AlphaRLee — APPROVED** [review]
